### PR TITLE
Make attributes lose their owner when removed

### PR DESCRIPTION
**Pull request by nox**
Test: https://github.com/w3c/web-platform-tests/pull/1718

### CODE REVIEW

File: components/script/dom/attr.rs [hunk 1]
```diff
@@ -6,8 +6,9 @@ use dom::bindings::cell::DOMRefCell;
 use dom::bindings::codegen::Bindings::AttrBinding::{self, AttrMethods};
 use dom::bindings::codegen::InheritTypes::NodeCast;
 use dom::bindings::global::GlobalRef;
-use dom::bindings::js::{JS, JSRef, Temporary};
-use dom::bindings::js::{OptionalRootedRootable, RootedReference};
+use dom::bindings::js::{JSRef, MutNullableJS, Temporary};
+use dom::bindings::js::{OptionalRootable, OptionalRootedRootable,
+                        RootedReference};
```
Comment (Manishearth) [inline components/script/dom/attr.rs]: nit: For multiline imports, we usually just import twice, eg

```
use dom::bindings::js::{OptionalRootable, OptionalRootedRootable};
use dom::bindings::js::RootedReference;
```

File: components/script/dom/attr.rs [hunk 2]
```diff
@@ -244,6 +247,24 @@ impl<'a> AttrHelpers<'a> for JSRef<'a, Attr> {
         &self.extended_deref().local_name
     }
 
+    fn set_owner(self, owner: Option<JSRef<Element>>) {
+        let ns = self.namespace.clone();
+        match (self.owner.get().root().r(), owner) {
+            (None, Some(new)) =>
```
Comment (Manishearth) [inline components/script/dom/attr.rs]: nit: braces on multiline match arms

File: components/script/dom/attr.rs [hunk 3]
```diff
@@ -244,6 +247,24 @@ impl<'a> AttrHelpers<'a> for JSRef<'a, Attr> {
...
+                // Already in the list of attributes of new owner.
+                assert!(new.get_attribute(ns, &self.local_name).root().r() == Some(self)),
```
Comment (Manishearth) [inline components/script/dom/attr.rs]: perhaps use a `debug_assert!`?

File: components/script/dom/attr.rs [hunk 4]
```diff
@@ -244,6 +247,24 @@ impl<'a> AttrHelpers<'a> for JSRef<'a, Attr> {
         &self.extended_deref().local_name
     }
 
+    fn set_owner(self, owner: Option<JSRef<Element>>) {
```
Comment (Manishearth) [inline components/script/dom/attr.rs]: I suggest you leave a doc comment on this stating that it should be used _after_ having moved the attr

**Comment by Manishearth** [comment]
Mostly looks good, save a few nits. r=me when they're fixed.


I assume there are no tests for this save for the one you upstreamed in wpt (which doesn't yet exist in our copy)

File: components/script/dom/attr.rs (hunk 3 above)
Comment (nox) [inline components/script/dom/attr.rs]: set_value() uses assert! for the same kind of invariant checking.

Comment (Manishearth) [inline components/script/dom/attr.rs]: okay, leave it then. We probably should move some asserts to debug asserts later, but for now I guess we can stick to asserts everywhere

**Comment by nox** [comment]
Nits fixed.

**Comment by Manishearth** [comment]
Rolled up into #5519 because of CI brokenness

**Comment by Ms2ger** [comment]
Rebased as https://github.com/servo/servo/pull/5524

**Pull request closed without merging**